src/python/gudhi/representations/vector_methods.py

```diff
@@ -161,53 +161,23 @@ def transform(self, X):
         Returns:
             numpy array with shape (number of diagrams) x (number of samples = **num_landscapes** x **resolution**): output persistence landscapes.
         """
-        num_diag, Xfit = len(X), []
-        x_values = np.linspace(self.sample_range[0], self.sample_range[1], self.new_resolution)
-        step_x = x_values[1] - x_values[0]
-
-        for i in range(num_diag):
-
-            diagram, num_pts_in_diag = X[i], X[i].shape[0]
-
-            ls = np.zeros([self.num_landscapes, self.new_resolution])
 
-            events = []
-            for j in range(self.new_resolution):
-                events.append([])
-
-            for j in range(num_pts_in_diag):
-                [px,py] = diagram[j,:2]
-                min_idx = np.clip(np.ceil((px          - self.sample_range[0]) / step_x).astype(int), 0, self.new_resolution)
-                mid_idx = np.clip(np.ceil((0.5*(py+px) - self.sample_range[0]) / step_x).astype(int), 0, self.new_resolution)
-                max_idx = np.clip(np.ceil((py          - self.sample_range[0]) / step_x).astype(int), 0, self.new_resolution)
-
-                if min_idx < self.new_resolution and max_idx > 0:
-
-                    landscape_value = self.sample_range[0] + min_idx * step_x - px
-                    for k in range(min_idx, mid_idx):
-                        events[k].append(landscape_value)
-                        landscape_value += step_x
-
-                    landscape_value = py - self.sample_range[0] - mid_idx * step_x
-                    for k in range(mid_idx, max_idx):
-                        events[k].append(landscape_value)
-                        landscape_value -= step_x
-
-            for j in range(self.new_resolution):
-                events[j].sort(reverse=True)
-                for k in range( min(self.num_landscapes, len(events[j])) ):
-                    ls[k,j] = events[j][k]
+        Xfit = []
+        x_values = np.linspace(self.sample_range[0], self.sample_range[1], self.new_resolution)
+        for i, diag in enumerate(X):
+            midpoints, heights = (diag[:, 0] + diag[:, 1]) / 2., (diag[:, 1] - diag[:, 0]) / 2.
+            tent_functions = np.maximum(heights[None, :] - np.abs(x_values[:, None] - midpoints[None, :]), 0)
+            tent_functions.partition(diag.shape[0] - self.num_landscapes, axis=1)
+            landscapes = np.sort(tent_functions, axis=1)[:, -self.num_landscapes:][:, ::-1].T
 
             if self.nan_in_range[0]:
-                ls = ls[:,1:]
+                landscapes = landscapes[:, 1:]
             if self.nan_in_range[1]:
-                ls = ls[:,:-1]
-            ls = np.sqrt(2)*np.reshape(ls,[1,-1])
-            Xfit.append(ls)
-
-        Xfit = np.concatenate(Xfit,0)
+                landscapes = landscapes[:, :-1]
```

Comment:
Couldn't we do that when we define `x_values`?
Also, Silhouette doesn't seem to skip those extreme elements. Should it?

Reply:
I extracted this logic and also applied it in Silhouettes.
FYI: the behavior is not the same in BettiCurves.

Reply:
If we change the behavior, we will need a note in the release notes to warn users about it.
This was mostly a question. @MathieuCarriere @gspr do you have an opinion which ones among Lanscape, Silhouette, BettiCurve, etc, should skip initial and final 0 value when they are asked to guess the bounds? And do we need to tweak the resolution so the output has the expected shape in that case?

Reply:
Without having thought very thoroughly about this, I think I'm preferential to `BettiCurve` *keeping* the initial and final 0. I think it makes the interpretation a bit more self-explanatory, as all values the curve takes are kept.

Reply:
I think it really depends what you want to do with the output. In order to plot it, having the extreme 0s is convenient. If you are going to feed the output to some classifier or other general ML method, having a coordinate that is always 0 is a waste, there is nothing to learn from it.

Reply:
> I think it really depends what you want to do with the output. In order to plot it, having the extreme 0s is convenient. If you are going to feed the output to some classifier or other general ML method, having a coordinate that is always 0 is a waste, there is nothing to learn from it.

True. But then again, the way these representations are documented (at least BettiCurve), they don't give any guarantees that for redundant information (such as repeated Betti numbers in a constant region for BettiCurve) is excluded. For BettiCurve, that happens to be the case, but I personally think that anyone feeding the output to some ML method should ask themselves whether it contains redundant information.

But again, I have no strong opinions on the matter, and don't think that removing the zeros is unreasonable either :-)

Reply:
> the way these representations are documented (at least BettiCurve), they don't give any guarantees

we can still aim for a good default.

Looking at the behavior in 3.6.0, when we specify only `resolution=10`, we have
* Landscape removes 0 on both sides
* Silhouette keeps 0 on both sides
* BettiCurve keeps the final 0 but not the initial one

I am sure we can at least improve consistency. My opinion is that we should drop 0 on both sides for all three. Note that this still leaves 0 on both sides for BettiCurve with resolution=None (@gspr, that seems to be the case you care about the most?), and may also leave some 0s on the side for the k-th landscape with k>1 (it isn't that bad to have a useless coordinate, I just want to avoid adding one that is too gratuitous). And we can add an option to bring back all the 0s for people who like plotting those curves, if there is demand.

@MathieuCarriere , would you be ok with such a plan? Anyone else?

(one thing I wondered while looking at it is, assuming the smallest interval that contains all the values where at least one function is non-zero is (0,12) and we want to sample 2 points in this interval, does it make more sense to pick 4 and 8 (as in 0, 4, 8, 12 minus the extremes) or 3 and 9 (a longer sequence would be -3, 3, 9, 15, so 0 and 12 fall in the middle of an interval)?)

Reply:
> BettiCurve with resolution=None (@gspr, that seems to be the case you care about the most?)

Yes, `resolution=None` was the use case that motivated me to write the current `BettiCurve`class.